libraries/c_sdk/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,3326 @@
+
+                    if( pNextHttpsRequest->isAsync && pNextHttpsRequest->pCallbacks->errorCallback )
+                    {
+                        pNextHttpsRequest->pCallbacks->errorCallback( pNextHttpsRequest->pUserPrivData, pNextHttpsRequest, NULL, scheduleStatus );
+                    }
+                    else
+                    {
+                        pNextHttpsRequest->pHttpsResponse->syncStatus = scheduleStatus;
+                    }
+                }
+            }
+        }
+        else
+        {
+            IotLogDebug( "Network receive callback found the request queue empty. A network send task was not scheduled." );
+        }
+    }
+
+    /* Dequeue response from the response queue now that it is finished. */
+    IotMutex_Lock( &( pHttpsConnection->respQMutex ) );
+
+    /* There could be a scenario where the request fails to send and the network server still responds,
+     * In this case, the failed response will have been cancelled and removed from the queue. If the network
+     * server still got a response, then the safest way to remove the current response is to remove it explicitly
+     * from the queue instead of dequeuing the header of the queue which might not be the current response. */
+    if( IotLink_IsLinked( &( pCurrentHttpsResponse->link ) ) )
+    {
+        IotDeQueue_Remove( &( pCurrentHttpsResponse->link ) );
+    }
+
+    IotMutex_Unlock( &( pHttpsConnection->respQMutex ) );
+
+    /* Signal to a synchronous reponse that the response is complete. */
+    if( pCurrentHttpsResponse->isAsync && pCurrentHttpsResponse->pCallbacks->responseCompleteCallback )
+    {
+        pCurrentHttpsResponse->pCallbacks->responseCompleteCallback( pCurrentHttpsResponse->pUserPrivData, pCurrentHttpsResponse, status, pCurrentHttpsResponse->status );
+    }
+
+    /* For a synchronous request release the semaphore. */
+    if( pCurrentHttpsResponse->isAsync == false )
+    {
+        IotSemaphore_Post( &( pCurrentHttpsResponse->respFinishedSem ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _createHttpsConnection( IotHttpsConnectionHandle_t * pConnHandle,
+                                                    IotHttpsConnectionInfo_t * pConnInfo )
+{
+    HTTPS_FUNCTION_ENTRY( IOT_HTTPS_OK );
+
+    IotNetworkError_t networkStatus = IOT_NETWORK_SUCCESS;
+
+    /* The maximum string length of the ALPN protocols is configured in IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH.
+     * The +1 is for the NULL terminator needed by IotNetworkCredentials_t.pAlpnProtos. */
+    char pAlpnProtos[ IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH + 1 ] = { 0 };
+
+    /* The maximum string length of the Server host name is configured in IOT_HTTPS_MAX_HOST_NAME_LENGTH.
+     * This +1 is for the NULL terminator needed by IotNetworkServerInfo_t.pHostName. */
+    char pHostName[ IOT_HTTPS_MAX_HOST_NAME_LENGTH + 1 ] = { 0 };
```

Comment:
What's the rationale for making this based on static configuration, and sticking it on the stack? If our copy isn't persisted beyond the scope of this function anyway, why not pass the caller's hostname pointer through to the network stack as-is?

Reply:
I create an array here because in IotHttpsConnectionInfo_t I do not necessitate that pAddress storing the host name needs a NULL terminator. There is an accompanying addressLen parameter. I also made the parameter a const as it must not change anyways.
```
 */
typedef struct IotHttpsConnectionInfo
{
    /**
     * @brief Remote server address that is DNS discoverable.
     *
     * For example: avs-alexa-na.amazon.com.
     */
    const char * pAddress;
    uint32_t addressLen; /**< @brief remote address length. */
    . . .
```
Because of this I cannot guarantee that the IotHttpsConnectionInfo_t.pAddress passed in by the user will be NULL terminated. 
The network interface structure that consumes the host name needs a NULL terminator, see iot_network.h:
```
typedef struct IotNetworkServerInfo
{
    const char * pHostName; /**< @brief Server host name. Must be NULL-terminated. */
    uint16_t port;          /**< @brief Server port in host-order. */
} IotNetworkServerInfo_t;
```

Because IotHttpsConnectionInfo_t.pAddress is a const I cannot simply append a NULL terminator, I need to allocate an array to append the NULL needed by the network interface. I did not want to use ANY mallocs in my implementation, so I created an array.

Please see: https://issues.amazon.com/issues/TS-8462